add-design (#252)
OWASP/common-requirement-enumeration: docs/designs/cre-links.md

# CRE Links

The biggest advantage of opencre.org is its ability to dynamically update itself by parsing other standards that link to it.

> So far, as a small project, our approach to this has been relatively simple: when any security resource added a link to opencre.org in their documents we would write a parser for their specific document format and information structure and add it to an ever-growing list of custom parsers.

Reviewer: I think this story misses the point that for many standards the information necessary to parse can be configured in the sources themselves, and for our generic parser to pick up. See further comments.

> So far, as a small project, our approach to this has been relatively simple: when any security resource added a link to opencre.org in their documents we would write a parser for their specific document format and information structure and add it to an ever-growing list of custom parsers.

Author: I added a suggested relevant solution at the end with an example

However, it fails for single-page documents with several sections or information and for mapping lots of small standards which will likely have their own unique structure each.

## The problem

Human-readable documents can have several structures and links can have even more, parsing and understanding human language is an open problem in computing.
For example, there are likely non-NLP/AI ways to extract the correct link and title from a markdown file structured like this.

``` Markdown
# Document Title

## Relevant Title
[...]

## References
https://opencre.org/cre/<>

```
We could make an assumption that we always pick the document title as the name for each document that links to us but then what happens if we get 2 cre links for different purposes? A markdown files could contain multiple sections of information.

In general parsing this kind of human readable information, complicates the project beyond its scope and maintaining those solutions requires significant resources.

= Suggested solution=

> Since standards writers are already required to provide links to opencre.org why not get them to encode all the information we need in the link?

Reviewer 2: ```suggestion

As standards integrate with the OpenCRE.org project, it would be intuitive for the authors to inject the information needed for the project to function with its maximum potential.
```

Reviewer: See earlier comment. The alternative is that this is done just once for a file or for a project, in a configuration comment, or in a separate configuration file for the opencre parser, containing standard name, version,  the grep patterns to search for the correct metadata and the pattern to construct the hyperlink.

> Then, not only does the parsing become simpler but we can also use analytics to discover resources that link to us but we may not know off yet.

Reviewer 2: ```suggestion
Then, not only does the parsing become simpler, but it also empowers us to use analytics to discover resources that link to us without our knowledge helping the ecosystem grow faster.
```

The link for a standard could be as simple as:

```
https://opencre.org/cre/<creid>?type=<standard>&name=<standard_name>&section=<section>&link=<where should we redirect to>
```

> The link for a tool rule could be as simple as:

Reviewer 2: ```suggestion
And a link for a tool rule would be:
```

> ```https://opencre.org/cre/<creid>?type=<tool>&name=<tool_name>&link=<where should we redirect to>&ruleID=<the tool rule id>

Reviewer: It's important to make clear in this story that these types of links are not the only solution. They are a nuisance to maintain, and are typically suitable for standards with inconsistent structure and/or with just a few links. For ASVS and cheat sheets for example this link structure should not be the way to go - unless they are generated from their own data.  Those standards are better parsed using a custom parser or a configured generic parser (following the early design for this). With such a parser there just needs to be a link to the cre topic, without any other parameters.

> ```https://opencre.org/cre/<creid>?type=<tool>&name=<tool_name>&link=<where should we redirect to>&ruleID=<the tool rule id>

Author: added a section describing that
